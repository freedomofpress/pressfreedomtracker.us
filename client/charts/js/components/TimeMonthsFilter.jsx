import React from 'react'
import * as d3 from 'd3'
import { sortBy, countBy } from 'lodash'
import { AnimatedDataset } from 'react-animated-dataset'
import { firstDayOfMonth } from '../lib/utilities'

const margins = {
	top: 15,
	left: 5,
	right: 30,
	bottom: 3,
}

function isDateValid(date) {
	return !Number.isNaN(new Date(date).getYear())
}

function monthWidth(date, scale) {
	return scale(d3.timeMonth.offset(new Date(date), 1)) - scale(new Date(date))
}

export default function TimeMonthsFilter({
	width,
	height,
	dateExtents,
	dataset,
	filterParameters: selectedDateRange,
	setFilterParameters: setSelectedDateRange,
}) {
	const [minDate, maxDate] = dateExtents

	function closeLineArea(dataset) {
		return [
			{
				date: minDate,
				count: 0,
			},
			...dataset,
			{
				date: maxDate,
				count: 0,
			},
		]
	}

	const monthFrequenciesObj = {
		...Object.fromEntries(
			d3.timeMonth.range(minDate, maxDate).map((date) => [date.toISOString(), 0])
		),
		...countBy(dataset, (d) => firstDayOfMonth(d.date).toISOString()),
	}
	const monthFrequencies = sortBy(
		Object.entries(monthFrequenciesObj).map(([dateISO, count]) => ({
			date: dateISO,
			count,
		})),
		'date'
	)

	const datasetSelected = dataset.filter(
		(d) =>
			new Date(d.date).getTime() >= selectedDateRange.min.getTime() &&
			new Date(d.date).getTime() <= selectedDateRange.max.getTime()
	)

	const monthFrequenciesSelectedObj = {
		...Object.fromEntries(
			d3.timeMonth.range(minDate, maxDate).map((date) => [date.toISOString(), 0])
		),
		...countBy(datasetSelected, (d) => firstDayOfMonth(d.date).toISOString()),
	}
	const monthFrequenciesSelected = sortBy(
		Object.entries(monthFrequenciesSelectedObj).map(([dateISO, count]) => ({
			date: dateISO,
			count,
		})),
		'date'
	)

	const xScale = d3
		.scaleTime()
		.domain(dateExtents)
		.range([1 + margins.left, width - margins.right])

	const yScale = d3
		.scaleLinear()
		.domain([0, d3.max(monthFrequencies.map((d) => d.count))])
		.range([0 + margins.top, height - margins.bottom - margins.top])
		.nice(3)

	const yScaleLine = yScale.copy().range([height - margins.bottom - margins.top, 0 + margins.top])

	const lineGenerator = d3
		.line()
		.curve(d3.curveStepAfter)
		.x((d) => xScale(new Date(d.date)))
		.y((d) => yScaleLine(d.count))

	return (
		<div style={{ flexDirection: 'row' }}>
			<svg width={width} height={height} style={{ fontFamily: 'var(--font-mono)' }}>
				<AnimatedDataset
					dataset={yScale.ticks(3)}
					tag="line"
					init={{
						opacity: 0,
					}}
					attrs={{
						x1: margins.left,
						x2: width,
						y1: (tick, i) => height - margins.bottom - yScale(tick, i) + (i === 0 ? 1.5 : 0),
						y2: (tick, i) => height - margins.bottom - yScale(tick) + (i === 0 ? 1.5 : 0),
						stroke: 'black',
						strokeWidth: (_, i) => (i === 0 ? 3 : 1),
						opacity: 1,
					}}
					keyFn={(d) => d}
					duration={450}
				/>
				<AnimatedDataset
					dataset={yScale.ticks(3)}
					tag="text"
					init={{
						opacity: 0,
					}}
					attrs={{
						x: width,
						y: (tick) => height - margins.bottom - yScale(tick) - 4,
						text: (tick) => tick,
						textAnchor: 'end',
						fontSize: 12,
						opacity: 1,
					}}
					keyFn={(d) => d}
					duration={450}
				/>
				<AnimatedDataset
					dataset={monthFrequencies}
					tag="rect"
					attrs={{
						x: (d) => xScale(new Date(d.date)),
						y: (d) => height - margins.bottom - yScale(d.count),
						fill: 'black',
						width: (d) => monthWidth(d.date, xScale) + 0.5,
						height: (d) => yScale(d.count) - margins.top,
						key: (_, i) => i,
					}}
					duration={300}
					keyFn={(d) => d.date}
				/>
				<AnimatedDataset
					dataset={monthFrequenciesSelected}
					tag="rect"
					attrs={{
						x: (d) => xScale(new Date(d.date)),
						y: (d) => height - margins.bottom - yScale(d.count),
						fill: '#F2FC67',
						width: (d) => monthWidth(d.date, xScale) + 0.5,
						height: (d) => yScale(d.count) - margins.top,
						key: (_, i) => i,
					}}
					duration={300}
					keyFn={(d) => d.date}
				/>
				<AnimatedDataset
					dataset={[monthFrequencies]}
					tag="path"
					attrs={{
						d: (d) => lineGenerator(sortBy(closeLineArea(monthFrequenciesSelected), (d) => d.date)),
						stroke: 'black',
						strokeWidth: 1,
						fill: 'none',
					}}
					duration={300}
					keyFn={(d, i) => i}
				/>
			</svg>

			<div
				style={{
					display: 'flex',
					width: '100%',
					justifyContent: 'space-between',
				}}
			>
				<input
					type={'date'}
					value={selectedDateRange.min.toISOString().slice(0, 10)}
					min={dateExtents[0].toISOString().slice(0, 10)}
					max={dateExtents[1].toISOString().slice(0, 10)}
					onChange={(event) => {
						if (isDateValid(event.target.value)) {
							const newDateRange = (oldDateRange) => {
								return { ...oldDateRange, min: new Date(event.target.value) }
							}
							setSelectedDateRange('filterTimeMonths', newDateRange)
						}
					}}
					style={{ flexBasis: '50%' }}
				/>
				<input
					type={'date'}
					value={selectedDateRange.max.toISOString().slice(0, 10)}
					min={dateExtents[0].toISOString().slice(0, 10)}
					max={dateExtents[1].toISOString().slice(0, 10)}
					onChange={(event) => {
						if (isDateValid(event.target.value)) {
							const newDateRange = (oldDateRange) => {
								return { ...oldDateRange, max: new Date(event.target.value) }
							}
							setSelectedDateRange('filterTimeMonths', newDateRange)
						}
					}}
					style={{ flexBasis: '50%' }}
				/>
			</div>
		</div>
	)
}
