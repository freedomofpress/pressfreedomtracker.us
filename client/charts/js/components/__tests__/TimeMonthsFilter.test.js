import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import TimeMonthsFilter from '../TimeMonthsFilter'
import * as d3 from 'd3'

test('renders TimeMonthsFilter with mocked data', () => {
	const renderer = new ShallowRenderer();
	expect(renderer.render(
		<TimeMonthsFilter
			width={500}
			height={500}
			dateExtents={[d3.timeParse('%m-%d-%Y')('1-1-2020'), d3.timeParse('%m-%d-%Y')('1-1-2022')]}
			dataset={[{ date: d3.timeParse('%m-%d-%Y')('6-1-2020') }]}
			filterParameters={{
				min: d3.timeParse('%m-%d-%Y')('1-1-2020'),
				max: d3.timeParse('%m-%d-%Y')('1-1-2022')
			}}
		/>
	)).toMatchSnapshot();
});
