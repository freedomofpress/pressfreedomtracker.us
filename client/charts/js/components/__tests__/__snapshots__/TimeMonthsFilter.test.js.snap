// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TimeMonthsFilter with mocked data 1`] = `
<div
  style={
    {
      "flexDirection": "row",
    }
  }
>
  <svg
    height={500}
    style={
      {
        "fontFamily": "var(--font-mono)",
      }
    }
    width={500}
  >
    <AnimatedDataset
      attrs={
        {
          "opacity": 1,
          "stroke": "black",
          "strokeWidth": [Function],
          "x1": 5,
          "x2": 500,
          "y1": [Function],
          "y2": [Function],
        }
      }
      dataset={
        [
          0,
          0.5,
          1,
        ]
      }
      duration={450}
      init={
        {
          "opacity": 0,
        }
      }
      keyFn={[Function]}
      tag="line"
    />
    <AnimatedDataset
      attrs={
        {
          "fontSize": 12,
          "opacity": 1,
          "text": [Function],
          "textAnchor": "end",
          "x": 500,
          "y": [Function],
        }
      }
      dataset={
        [
          0,
          0.5,
          1,
        ]
      }
      duration={450}
      init={
        {
          "opacity": 0,
        }
      }
      keyFn={[Function]}
      tag="text"
    />
    <AnimatedDataset
      attrs={
        {
          "fill": "black",
          "height": [Function],
          "key": [Function],
          "width": [Function],
          "x": [Function],
          "y": [Function],
        }
      }
      dataset={
        [
          {
            "count": 0,
            "date": "2020-01-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-02-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-03-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-04-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-05-01T00:00:00.000Z",
          },
          {
            "count": 1,
            "date": "2020-06-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-07-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-08-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-09-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-10-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-11-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-12-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-01-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-02-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-03-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-04-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-05-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-06-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-07-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-08-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-09-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-10-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-11-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-12-01T00:00:00.000Z",
          },
        ]
      }
      duration={300}
      keyFn={[Function]}
      tag="rect"
    />
    <AnimatedDataset
      attrs={
        {
          "fill": "#F2FC67",
          "height": [Function],
          "key": [Function],
          "width": [Function],
          "x": [Function],
          "y": [Function],
        }
      }
      dataset={
        [
          {
            "count": 0,
            "date": "2020-01-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-02-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-03-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-04-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-05-01T00:00:00.000Z",
          },
          {
            "count": 1,
            "date": "2020-06-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-07-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-08-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-09-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-10-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-11-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2020-12-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-01-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-02-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-03-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-04-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-05-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-06-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-07-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-08-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-09-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-10-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-11-01T00:00:00.000Z",
          },
          {
            "count": 0,
            "date": "2021-12-01T00:00:00.000Z",
          },
        ]
      }
      duration={300}
      keyFn={[Function]}
      tag="rect"
    />
    <AnimatedDataset
      attrs={
        {
          "d": [Function],
          "fill": "none",
          "stroke": "black",
          "strokeWidth": 1,
        }
      }
      dataset={
        [
          [
            {
              "count": 0,
              "date": "2020-01-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2020-02-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2020-03-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2020-04-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2020-05-01T00:00:00.000Z",
            },
            {
              "count": 1,
              "date": "2020-06-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2020-07-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2020-08-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2020-09-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2020-10-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2020-11-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2020-12-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2021-01-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2021-02-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2021-03-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2021-04-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2021-05-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2021-06-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2021-07-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2021-08-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2021-09-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2021-10-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2021-11-01T00:00:00.000Z",
            },
            {
              "count": 0,
              "date": "2021-12-01T00:00:00.000Z",
            },
          ],
        ]
      }
      duration={300}
      keyFn={[Function]}
      tag="path"
    />
  </svg>
  <div
    style={
      {
        "display": "flex",
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
  >
    <input
      max="2022-01-01"
      min="2020-01-01"
      onChange={[Function]}
      style={
        {
          "flexBasis": "50%",
        }
      }
      type="date"
      value="2020-01-01"
    />
    <input
      max="2022-01-01"
      min="2020-01-01"
      onChange={[Function]}
      style={
        {
          "flexBasis": "50%",
        }
      }
      type="date"
      value="2022-01-01"
    />
  </div>
</div>
`;
